models/sql_postgres_with_schema.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"sync"
+
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/lib/pq"
+	"xorm.io/xorm/dialects"
+)
+
+var registerOnce sync.Once
+
+func registerPostgresSchemaDriver() {
+	registerOnce.Do(func() {
+		sql.Register("postgresschema", &postgresSchemaDriver{})
+		dialects.RegisterDriver("postgresschema", dialects.QueryDriver("postgres"))
+	})
+}
+
+type postgresSchemaDriver struct {
+	pq.Driver
+}
+
+// Open opens a new connection to the database. name is a connection string.
+// This function opens the postgres connection in the default manner but immediately
+// runs set_config to set the search_path appropriately
+func (d *postgresSchemaDriver) Open(name string) (driver.Conn, error) {
+	conn, err := d.Driver.Open(name)
+	if err != nil {
+		return conn, err
+	}
+
+	stmt, err := conn.Prepare(`SELECT set_config(
```

Comment:
We need a map to know if the connection has been set. The open may get the connection from pool or create a new connection. 

And why we need `Prepare` here? I think just `Exec` is OK.

Reply:
https://golang.org/pkg/database/sql/driver/#Conn driver.Conn does not have Exec.

the pq driver does not get from a pool.

Reply:
Yeah. For my first question. You are right. This is a low-level Open, the connection pool is after this one. So always execute the set path is OK.

For the second one, you can always use an assertion to know if it's a `driver.Execer` because the driver has implemented that. https://github.com/lib/pq/blob/master/conn.go#L884. And for more common, we can also detect if it's implemented `ExecerContext` at first if the driver implemented that in future.

```
execerCtx, ok := conn.(driver.ExecerContext)

var execer driver.Execer
if !ok {
  execer, ok = conn.(driver.Execer)
}
```

Reply:
pq is in maintenance mode only so as far as I can see it will never implement execercontext. in any case if we want to use the context based things we should implement DriverContext and pass the context down through that.